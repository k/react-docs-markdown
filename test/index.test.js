/* eslint-disable global-require */
import fs from 'fs';
import path from 'path';
import docsToMarkdown from '../src/index';

const inputJson = require('./input/CaseList.json');
const imageJson = require('./input/Image.json');
const textAreaJson = require('./input/TextArea.json');

describe('Docs to markdown', () => {
  it('should convert caselist', () => {
    const result = docsToMarkdown(inputJson, 'CaseList');
    expect(result).toMatchSnapshot();
    expect(typeof result).toEqual('string');
    expect(result.includes('### margin')).toEqual(true);
    expect(result.includes('Arrayof')).toEqual(true);
    expect(result.includes('@custom-exclude')).toEqual(true);
    expect(result.includes('@internal')).toEqual(false); // The @internal description should be excluded.
    save('CaseList', result);
  });

  it('should convert Image', () => {
    const result = docsToMarkdown(imageJson, 'Image');
    expect(result).toMatchSnapshot();
    save('Image', result);
  });

  it('should convert TextArea', () => {
    const result = docsToMarkdown(textAreaJson, 'TextArea');
    expect(result).toMatchSnapshot();
    save('TextArea', result);
  });

  it('should handle no props', () => {
    const result = docsToMarkdown({description: 'Short description', props: null}, 'CaseList');
    expect(result).toMatchSnapshot();
  });

  describe('Exclude key', () => {
    it('should exclude key with string', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeKeys: 'margin'});
      expect(result).toMatchSnapshot();
      expect(result.includes('### margin')).toEqual(false);
    });

    it('should exclude key with array', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeKeys: ['margin']});
      expect(result).toMatchSnapshot();
      expect(result.includes('### margin')).toEqual(false);
    });

    it('should exclude key with regex', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeKeys: /margin/});
      expect(result).toMatchSnapshot();
      expect(result.includes('### margin')).toEqual(false);
    });
  });

  describe('Exclude type', () => {
    it('should exclude type with string', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeTypes: 'arrayOf'});
      expect(result).toMatchSnapshot();
      expect(result.includes('Arrayof')).toEqual(false);
    });

    it('should exclude type with array', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeTypes: ['arrayOf']});
      expect(result).toMatchSnapshot();
      expect(result.includes('Arrayof')).toEqual(false);
    });

    it('should exclude type with regex', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeTypes: /arrayOf/});
      expect(result).toMatchSnapshot();
      expect(result.includes('Arrayof')).toEqual(false);
    });
  });

  describe('Exclude description', () => {
    it('should exclude type with string', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeDescription: '@custom-exclude'});
      expect(result).toMatchSnapshot();
      expect(result.includes('@custom-exclude')).toEqual(false);
      // Should remove the default excludeDescription, so
      expect(result.includes('@internal')).toEqual(true);
    });

    it('should exclude type with array', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeDescription: ['@custom-exclude']});
      expect(result).toMatchSnapshot();
      expect(result.includes('@custom-exclude')).toEqual(false);
    });

    it('should exclude type with regex', () => {
      const result = docsToMarkdown(inputJson, 'CaseList', {excludeDescription: /@custom-exclude/});
      expect(result).toMatchSnapshot();
      expect(result.includes('@custom-exclude')).toEqual(false);
    });
  });
});

function save(name, md) {
  fs.writeFile(path.join(__dirname, `output/${name}.md`), md, (err) => {
    if (err) throw err;
  });
}
