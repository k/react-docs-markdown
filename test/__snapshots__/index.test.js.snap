// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Docs to markdown Exclude description should exclude type with array 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type    | Default | Required | Description                                                        |
| --------------------------------- | ------- | ------- | -------- | ------------------------------------------------------------------ |
| [items](#markdown-header-items)   | Arrayof |         | true     |                                                                    |
| [margin](#markdown-header-margin) | OneOf   | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| component                         | Node    |         | false    | @internal Use another root component.                              |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'

--------------------------------------------------------------------------------

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown Exclude description should exclude type with regex 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type    | Default | Required | Description                                                        |
| --------------------------------- | ------- | ------- | -------- | ------------------------------------------------------------------ |
| [items](#markdown-header-items)   | Arrayof |         | true     |                                                                    |
| [margin](#markdown-header-margin) | OneOf   | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| component                         | Node    |         | false    | @internal Use another root component.                              |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'

--------------------------------------------------------------------------------

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown Exclude description should exclude type with string 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type    | Default | Required | Description                                                        |
| --------------------------------- | ------- | ------- | -------- | ------------------------------------------------------------------ |
| [items](#markdown-header-items)   | Arrayof |         | true     |                                                                    |
| [margin](#markdown-header-margin) | OneOf   | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| component                         | Node    |         | false    | @internal Use another root component.                              |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'

--------------------------------------------------------------------------------

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown Exclude key should exclude key with array 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                            | Type    | Default | Required | Description     |
| ------------------------------- | ------- | ------- | -------- | --------------- |
| [items](#markdown-header-items) | Arrayof |         | true     |                 |
| backgroundColor                 | String  |         | true     | @custom-exclude |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'
"
`;

exports[`Docs to markdown Exclude key should exclude key with regex 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                            | Type    | Default | Required | Description     |
| ------------------------------- | ------- | ------- | -------- | --------------- |
| [items](#markdown-header-items) | Arrayof |         | true     |                 |
| backgroundColor                 | String  |         | true     | @custom-exclude |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'
"
`;

exports[`Docs to markdown Exclude key should exclude key with string 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                            | Type    | Default | Required | Description     |
| ------------------------------- | ------- | ------- | -------- | --------------- |
| [items](#markdown-header-items) | Arrayof |         | true     |                 |
| backgroundColor                 | String  |         | true     | @custom-exclude |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'
"
`;

exports[`Docs to markdown Exclude type should exclude type with array 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type   | Default | Required | Description                                                        |
| --------------------------------- | ------ | ------- | -------- | ------------------------------------------------------------------ |
| [margin](#markdown-header-margin) | OneOf  | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| backgroundColor                   | String |         | true     | @custom-exclude                                                    |

## Complex Props

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown Exclude type should exclude type with regex 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type   | Default | Required | Description                                                        |
| --------------------------------- | ------ | ------- | -------- | ------------------------------------------------------------------ |
| [margin](#markdown-header-margin) | OneOf  | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| backgroundColor                   | String |         | true     | @custom-exclude                                                    |

## Complex Props

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown Exclude type should exclude type with string 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type   | Default | Required | Description                                                        |
| --------------------------------- | ------ | ------- | -------- | ------------------------------------------------------------------ |
| [margin](#markdown-header-margin) | OneOf  | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| backgroundColor                   | String |         | true     | @custom-exclude                                                    |

## Complex Props

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown should convert Image 1`] = `
"# Image

Image handles fetching the correct Imgix image, and rendering it onto the page.
By default it will load an image that fills the width of it's container, while keeping the original image aspect ratio


## Props
| Name                               | Type   | Default | Required | Description                                                                                                                   |
| ---------------------------------- | ------ | ------- | -------- | ----------------------------------------------------------------------------------------------------------------------------- |
| children                           | Node   |         | false    |                                                                                                                               |
| src                                | String |         | true     | Image src URL                                                                                                                 |
| height                             | Number |         | true     | Original image height                                                                                                         |
| width                              | Number |         | true     | Original image width                                                                                                          |
| alt                                | String |         | false    | Image alt text                                                                                                                |
| contain                            | Bool   | false   | false    | Contain the image inside the area, instead of cropping                                                                        |
| [alignX](#markdown-header-align-x) | Enum   |         | false    | Align the image on X axis, default to center in the browser                                                                   |
| [alignY](#markdown-header-align-y) | Enum   |         | false    | Align the image on Y axis, default to center in the browser                                                                   |
| [crop](#markdown-header-crop)      | Enum   | 'faces' | false    | The method used when cropping the image: https://docs.imgix.com/apis/url/size/crop                                            |
| padded                             | Bool   | false   | false    | Add padding around the image - use with \\"contain\\" to create a padded area around the image, so it doesn't extend to the edge. |
| fill                               | Bool   | false   | false    | Fill the container element. On mobile it will revert to aspect ratio                                                          |
| mobileFill                         | Bool   | false   | false    | Also apply the fill on mobile devices.                                                                                        |
| mobileWidthScale                   | Number |         | false    | On mobile, scale the width of the aspect ratio with this - Use to crop the width                                              |
| maxAspectRatio                     | Number | 2       | false    | The maximum aspect ratio to use after (height / width) - Adds some sanity to editor images, to ensure they are not too tall.  |
| imgixParams                        | Object |         | false    | Any custom Imgix params                                                                                                       |
| imageProps                         | Object |         | false    | Extra tags to pass as img HTML attributes                                                                                     |
| fadeIn                             | Bool   | false   | false    | Fade in the image when it's loaded. Creates a nice effect with the blurred image                                              |
| disableBlur                        | Bool   | false   | false    | Disable the blurred placeholder shown until size is ready. Will instead show a spinner                                        |

## Complex Props

### alignX
Type: _Enum_

Align the image on X axis, default to center in the browser

**alignX** should be one of the following values:

* 'left'
* 'right'
* 'center'

--------------------------------------------------------------------------------

### alignY
Type: _Enum_

Align the image on Y axis, default to center in the browser

**alignY** should be one of the following values:

* 'top'
* 'bottom'
* 'center'

--------------------------------------------------------------------------------

### crop
Type: _Enum_

The method used when cropping the image: https://docs.imgix.com/apis/url/size/crop

**crop** should be one of the following values:

* 'faces'
* 'entropy'
* 'edges'
* 'top'
* 'bottom'
* 'left'
* 'right'
* \\"\\"
* null
"
`;

exports[`Docs to markdown should convert TextArea 1`] = `
"# TextArea

## Props
| Name     | Type   | Default                               | Required | Description |
| -------- | ------ | ------------------------------------- | -------- | ----------- |
| debounce | Object | {<br>  blur: 0,<br>  change: 500<br>} | false    |             |
| updateOn | String | 'blur change'                         | false    |             |
| value    | String |                                       | false    |             |
| onBlur   | Func   | () => {}                              | false    |             |
"
`;

exports[`Docs to markdown should convert caselist 1`] = `
"# CaseList

Overview of all the components


## Props
| Name                              | Type    | Default | Required | Description                                                        |
| --------------------------------- | ------- | ------- | -------- | ------------------------------------------------------------------ |
| [items](#markdown-header-items)   | Arrayof |         | true     |                                                                    |
| [margin](#markdown-header-margin) | OneOf   | 'small' | false    | Margin can be either a string, or an object with top/bottom margin |
| backgroundColor                   | String  |         | true     | @custom-exclude                                                    |

## Complex Props

### items
Type: _Arrayof_

**items** is an array of the following type:

| Name                                | Type   | Required |
| ----------------------------------- | ------ | -------- |
| className                           | String | false    |
| backgroundColor                     | String | false    |
| backgroundImage                     | String | false    |
| color                               | String | false    |
| fullWidth                           | Bool   | false    |
| category                            | String | false    |
| title                               | String | true     |
| body                                | String | false    |
| [link](#markdown-header-link)       | Object | true     |
| [preview](#markdown-header-preview) | Object | false    |

#### link
Type: _Object_

| Name   | Type   | Required |
| ------ | ------ | -------- |
| href   | String | true     |
| target | String | false    |

#### preview
Type: _Object_

| Name                          | Type   | Required |
| ----------------------------- | ------ | -------- |
| image                         | String | true     |
| [type](#markdown-header-type) | Enum   | true     |

##### type
Type: _Enum_

* 'iphone'
* 'ipad'
* 'desktop'
* 'dual'

--------------------------------------------------------------------------------

### margin
Type: _OneOf_

Margin can be either a string, or an object with top/bottom margin

**margin** should be one of the following types:

#### Object

> | Name                              | Type | Required |
> | --------------------------------- | ---- | -------- |
> | [top](#markdown-header-top)       | Enum | false    |
> | [bottom](#markdown-header-bottom) | Enum | false    |
> 
> ##### top
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
> ##### bottom
> Type: _Enum_
> 
> * 'none'
> * 'small'
> * 'large'
> 
#### Enum

> * 'none'
> * 'small'
> * 'large'
> "
`;

exports[`Docs to markdown should handle no props 1`] = `
"# CaseList

Short description

## No props"
`;
